homeassistant/components/http/forwarded.py

```diff
@@ -0,0 +1,165 @@
+        connected_ip = ip_address(request.transport.get_extra_info("peername")[0])
+        if not any(connected_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+            _LOGGER.warning(
+                "Received X-Forwarded-For header from untrusted proxy %s, headers not processed",
+                str(connected_ip),
+            )
+            # Not trusted, continue as normal
+            return await handler(request)
+
+        # Multiple X-Forwarded-For headers
+        if len(forwarded_for) > 1:
+            _LOGGER.error("Too many headers for X-Forwarded-For: %s", forwarded_for)
+            raise HTTPBadRequest
+
+        # Process X-Forwarded-For from the right side (by reversing the list)
+        forwarded_for = list(reversed(forwarded_for[0].split(",")))
+        try:
+            forwarded_for = [ip_address(addr.strip()) for addr in forwarded_for]
+        except ValueError:
+            _LOGGER.error("Invalid IP address in X-Forwarded-For: %s", forwarded_for)
+            raise HTTPBadRequest
+
+        # Find the last trusted index in the X-Forwarded-For list
+        index = 0
+        for forwarded_ip in forwarded_for:
+            if any(forwarded_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+                index += 1
+                continue
+            overrides["remote"] = str(forwarded_ip)
+            break
+
+        # If all the IP addresses are from trusted networks, take the left-most.
+        if "remote" not in overrides:
+            index = -1
+            overrides["remote"] = str(forwarded_for[-1])
+
+        # Handle X-Forwarded-Proto
+        forwarded_proto = list(reversed(request.headers.getall(X_FORWARDED_PROTO, [])))
+        if forwarded_proto:
+            if len(forwarded_proto) > 1:
+                _LOGGER.error(
+                    "Too many headers for X-Forward-Proto: %s", forwarded_proto
+                )
+                raise HTTPBadRequest
+
+            forwarded_proto = [
+                proto
+                for proto in (p.strip() for p in forwarded_proto[0].split(","))
+                if proto
+            ]
+
+            # The X-Forwarded-Proto contains either one element, or the equals number
+            # of elements as X-Forwarded-For
+            if len(forwarded_proto) != 1 and len(forwarded_proto) != len(forwarded_for):
+                _LOGGER.error(
+                    "Incorrect number of elements in X-Forward-Proto. Expected 1 or %d, got %d: %s",
+                    len(forwarded_for),
+                    len(forwarded_proto),
+                    forwarded_proto,
+                )
+                raise HTTPBadRequest
+
+            # Ideally this should take the scheme corresponding to the entry
+            # in X-Forwarded-For that was chosen, but some proxies only retain
+            # one element. In that case, use what we have.
+            if index != -1 and index >= len(forwarded_proto):
```

Comment:
I think that we should be more explicit here. We get here and either have a single protocol or we have as many protocols as IPs. 

Should we rewrite this check to be something like:

```python
if index > 1 and len(forwarded_proto) == 1:
    proto_index = -1
else:
    proto_index = index
```

it is a bit weird though that we override the index value that was used for the IP. Should we also make this a new variable?

Reply:
This part is actually missing tests, it is covered, but not explicitly.

Reply:
Alright, it contained a bug actually. Protected by tests now. Implemented something similar to the suggestion.